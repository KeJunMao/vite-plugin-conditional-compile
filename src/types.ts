import type { FilterPattern } from "vite";

export interface Options {
  /**
   * @default ["**\/*"]
   */
  include: FilterPattern;
  /**
   * @default []
   */
  exclude: FilterPattern;
}

export type UserOptions = Partial<Options>;

export type ResolvedOptions = Options;
