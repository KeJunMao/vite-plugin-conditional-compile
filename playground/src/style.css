body {
  /* #v-ifdef DEV */
  color: red;
  /* #v-else */
  color: blue;
  /* #v-endif */
}
