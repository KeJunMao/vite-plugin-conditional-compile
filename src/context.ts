import MagicString from "magic-string";
import { createFilter, createLogger, Logger } from "vite";
import { ResolvedOptions } from "./types";

export class Context {
  env!: Record<string, any>;
  blockRegex: RegExp;
  filter: any;
  logger: Logger;
  constructor(public options: ResolvedOptions) {
    this.blockRegex =
      /^.*?#v-if(n?)def\s*(\S+).*[\r\n]{1,2}([\s\S]+?)\s*.*?#v-endif.*?$/gm;
    this.filter = createFilter(this.options.include, this.options.exclude);
    this.logger = createLogger("info", {
      prefix: "[vite-plugin-conditional-compile]",
    });
  }

  setEnv(env: Record<string, any>) {
    this.env = env;
  }

  transform(code: string, id: string) {
    try {
      if (!this.filter(id)) return;
      const source = this.createSource(code, id);
      source.replace(this.blockRegex, this.blockReplaceHandler.bind(this));

      if (source.hasChanged()) {
        return {
          code: source.toString(),
          map: source.generateMap({
            source: id,
            file: `${id}.map`,
            includeContent: true,
          }),
        };
      }
    } catch (error: any) {
      this.logger.error(error, {
        timestamp: true,
      });
    }
  }

  blockReplaceHandler(_: string, ...args: any[]) {
    const {
      isNot: startNot,
      conditional,
      code,
    } = this.parseBlockReplaceHandlerParams(args);
    const conditionals = this.parseConditional(conditional);
    const isKeep = conditionals.some((subConditional) => {
      const isNot = startNot !== subConditional.isNot;
      const value = this.env[subConditional.key];
      if (!Object.hasOwn(this.env, subConditional.key)) {
        this.logger.warn(
          `No '${subConditional.key}' propertie in Vite environment variables`,
          {
            timestamp: true,
          }
        );
      }
      if (subConditional.value === undefined) {
        return !!value !== isNot;
      } else {
        return (String(value) === subConditional.value) !== isNot;
      }
    });
    const { ifCode, elseCode } = this.parseElse(code);
    return isKeep ? ifCode : elseCode;
  }

  parseBlockReplaceHandlerParams([$1, $2, $3]: string[]) {
    return {
      isNot: !!$1,
      conditional: $2,
      code: $3,
    };
  }

  parseConditional(conditional: string) {
    return conditional.split("||").map((subConditional) => {
      // @ts-expect-error
      const [_, startNot, key, isNot, value] = subConditional.match(
        /^(!?)(\S+?)(?:(!?)(?:=)(\S+))?$/
      );
      return {
        key,
        isNot: !!startNot !== !!isNot,
        value,
      };
    });
  }

  parseElse(code: string) {
    const [_, ifCode, elseCode] = code?.match(
      /^([\s\S]*?).*#v-else.*[\r\n]{1,2}([\s\S]*)$/
    ) ?? [code, "", ""];
    return {
      ifCode: ifCode ? ifCode : code,
      elseCode: elseCode ? elseCode : "",
    };
  }

  createSource(code: string, id: string) {
    return new MagicString(code, {
      filename: id,
    });
  }
}
