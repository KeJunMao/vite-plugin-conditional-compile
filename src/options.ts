import { ResolvedOptions, UserOptions } from "./types";

export const resolveOptions = (userOptions?: UserOptions): ResolvedOptions => {
  return {
    include: ["**/*"],
    exclude: [],
    ...userOptions,
  };
};
