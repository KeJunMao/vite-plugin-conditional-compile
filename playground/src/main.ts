import "./style.css";

const render = (s: string) => {
  // @ts-expect-error
  const p = document.createElement("p");
  p.innerText = s;
  // @ts-expect-error
  document.body.appendChild(p);
};

render("no conditional");

// #v-ifdef DEV
render("Conditional: DEV");
// #v-endif

// #v-ifdef PROD
render("Conditional: PROD");
// #v-endif

// #v-ifdef !DEV
render("Conditional: !DEV");
// #v-endif

// #v-ifdef !PROD
render("Conditional: !PROD");
// #v-endif

// #v-ifdef DEV
render("Conditional: n DEV");
// #v-endif

// #v-ifdef PROD
render("Conditional: n PROD");
// #v-endif

// #v-ifdef (DEV||PROD)
render("Conditional: DEV||PROD");
// #v-endif

// #v-ifdef (!DEV||PROD)
render("Conditional: !DEV||PROD");
// #v-endif

// #v-ifdef DEV=true
render("Conditional: DEV=true");
// #v-endif

// #v-ifdef PROD!=true
render("Conditional: PROD!=true");
// #v-endif

// #v-ifdef !DEV=false
render("Conditional: !DEV=false");
// #v-endif

// #v-ifdef !DEV!=true
render("Conditional: !DEV!=true");
// #v-endif

// #v-ifdef (!DEV=true||PROD=true)
render("Conditional: !DEV=true||PROD=true");
// #v-else
render("Conditional: !DEV=true||PROD=true else");
// #v-endif

// #v-ifdef (DEV!=true||PROD=true)
render("Conditional: n DEV!=true||PROD=true");
// #v-endif
