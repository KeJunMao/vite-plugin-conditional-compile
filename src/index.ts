import type { Plugin } from "vite";
import { UserOptions } from "./types";
import { Context } from "./context";
import { resolveOptions } from "./options";

const VitePluginConditionalCompile = (
  userOptions: UserOptions = {}
): Plugin => {
  const options = resolveOptions(userOptions);
  const ctx = new Context(options);
  return {
    name: "vite-plugin-conditional-compile",
    enforce: "pre",
    configResolved(_config) {
      ctx.setEnv(_config.env);
    },
    transform: (code, id) => ctx.transform(code, id),
  };
};

export default VitePluginConditionalCompile;
