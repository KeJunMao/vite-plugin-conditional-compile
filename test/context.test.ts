import { describe, it } from "vitest";
import { createContext } from "../src/context";
import { resolveOptions } from "../src/options";
import { resolve } from "node:path";
import fs from "fs-extra";
import fg from "fast-glob";

async function getGlobContent(cwd: string, glob: string) {
  return await fg(glob, { cwd, absolute: true })
    .then((r) => Promise.all(r.map((f) => fs.readFile(f, "utf8"))))
    .then((r) => r.join("\n"));
}

async function createCtxWithEnv(env: Record<string, any> = {}) {
  const opts = resolveOptions();
  const ctx = createContext(opts);
  ctx.env = env
  return ctx;
}

describe("Context", async () => {
  const root = resolve(__dirname, "fixtures/code");
  const sourceCode = await getGlobContent(root, "*.*");
  it("empty env", async ({ expect }) => {
    const ctx = await createCtxWithEnv();
    const transformCode = ctx.transform(sourceCode, "fixtures");
    expect(transformCode).not.to.equal(undefined);
    expect(transformCode).toMatchSnapshot();
  });
  it("DEV env", async ({ expect }) => {
    const ctx = await createCtxWithEnv({
      DEV: true,
    });
    const transformCode = ctx.transform(sourceCode, "fixtures");
    expect(transformCode).not.to.equal(undefined);
    expect(transformCode).toMatchSnapshot();
  });
  it("PROD env", async ({ expect }) => {
    const ctx = await createCtxWithEnv({
      PROD: true,
    });
    const transformCode = ctx.transform(sourceCode, "fixtures");
    expect(transformCode).not.to.equal(undefined);
    expect(transformCode).toMatchSnapshot();
  });
});
