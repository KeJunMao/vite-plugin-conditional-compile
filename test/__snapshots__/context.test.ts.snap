// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Context > DEV env 1`] = `
"body {
  content: \\"DEV\\";

}

{
  \\"data\\": \\"PROD else\\"
}

import { test } from \\"vitest\\";
console.log(\\"DEV\\", \\"vitest\\");


const Component = () => <div>
    DEV
    DEV=true

</div>;

<template>
  <div class=\\"app\\">

    Default
  </div>
</template>

<script lang=\\"ts\\" setup>
const data = ref(0);

</script>

<style>
.app {
  color: red;
}
</style>
"
`;

exports[`Context > PROD env 1`] = `
"body {
  content: \\"DEV else\\";
}

{
  \\"data\\": \\"PROD\\",

}

import { test } from \\"node:test\\";
console.log(\\"DEV else\\", \\"node:test\\");

const Component = () => <div>


    !PROD!=true
</div>;

<template>
  <div class=\\"app\\">
    !DEV
    Default
  </div>
</template>

<script lang=\\"ts\\" setup>
const data = ref(0);
data.value = 1;
</script>

<style>
.app {

}
</style>
"
`;

exports[`Context > empty env 1`] = `
"body {
  content: \\"DEV else\\";
}

{
  \\"data\\": \\"PROD else\\"
}

import { test } from \\"node:test\\";
console.log(\\"DEV else\\", \\"node:test\\");

const Component = () => <div>



</div>;

<template>
  <div class=\\"app\\">
    !DEV
    Default
  </div>
</template>

<script lang=\\"ts\\" setup>
const data = ref(0);

</script>

<style>
.app {

}
</style>
"
`;
